core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -278,10 +278,11 @@ class PartitionStateMachine(controller: KafkaController) extends Logging {
         throw new StateChangeFailedException(failMsg)
       case _ =>
         debug("Live assigned replicas for partition %s are: [%s]".format(topicAndPartition, liveAssignedReplicas))
-        // make the first replica in the list of assigned replicas, the leader
-        val leader = liveAssignedReplicas.head
-        val leaderIsrAndControllerEpoch = new LeaderIsrAndControllerEpoch(new LeaderAndIsr(leader, liveAssignedReplicas.toList),
-          controller.epoch)
+        val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(
+          // leader is the first replica in the list of assigned replicas
+          LeaderAndIsr(liveAssignedReplicas.head, liveAssignedReplicas.toList),
```

Comment:
The existing code is pretty odd. Since you're touching this code, maybe you could improve it a little. Instead of checking if the size is `0` via a pattern match, we could do `liveAssignedReplicas.headOption` and go from there.